### PR TITLE
refactoring sqlite persistence plugin

### LINKED ISSUE #289: Sqlite persistence is in fact EntityFramework persistence, rename?
Sqlite persistence is in fact EntityFramework persistence, may be it is a good idea to name it like this so people can use the database of their choice?

### PR DESCRIPTION
**Pull request by raskolnikoov**
refactoring sqlite persistence plugin to use microsoft.data.sqlite instead of EF sqlite.

closes #289

### CODE REVIEW

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 1]
```diff
@@ -1,22 +1,41 @@
...
+using System.Linq;
 
 namespace Proto.Persistence.Sqlite
 {
     public class SqliteProvider : IProvider
     {
         private readonly string _datasource;
+        private string _connectionString => $"{new SqliteConnectionStringBuilder { DataSource = _datasource }}";
 
         public SqliteProvider(string datasource = "states.db")
```
Comment (adamhathcock) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Maybe add an overload to accept a SqliteConnectionStringBuilder as providing a custom string with different sqlite settings is something I'd want to do

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 2]
```diff
@@ -73,18 +102,35 @@ public SqliteProvider(string datasource = "states.db")
...
+                using (var connection = new SqliteConnection(_connectionString))
                 {
-                    var items = db.Events
-                        .Where(x => x.ActorName == actorName)
-                        .Where(x => x.EventIndex >= indexStart && x.EventIndex <= indexEnd)
-                        .ToList();
+                    connection.Open();
 
-                    foreach (var item in items)
+                    using (var transaction = connection.BeginTransaction())
```
Comment (adamhathcock) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Don't need a transaction over a single cursor read.

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 3]
```diff
@@ -29,19 +48,24 @@ public SqliteProvider(string datasource = "states.db")
...
-                    var items = db.Events
-                                    .Where(x => x.ActorName == actorName)
-                                    .Where(x => x.EventIndex <= inclusiveToIndex)
-                                    .ToList();
+                    connection.Open();
 
-                    db.RemoveRange(items);
-
-                    await db.SaveChangesAsync();
+                    using (var transaction = connection.BeginTransaction())
```
Comment (adamhathcock) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Don't need a transaction over a single command.  Sqlite implicitly has a transaction on this.

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 4]
```diff
@@ -51,16 +75,21 @@ public SqliteProvider(string datasource = "states.db")
...
-                    var items = db.Snapshots
-                                    .Where(x => x.ActorName == actorName)
-                                    .Where(x => x.SnapshotIndex <= inclusiveToIndex)
-                                    .ToList();
+                    connection.Open();
 
-                    db.RemoveRange(items);
-
-                    await db.SaveChangesAsync();
+                    using (var transaction = connection.BeginTransaction())
```
Comment (adamhathcock) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Don't need a transaction over a single command.  Sqlite implicitly has a transaction on this.

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 5]
```diff
@@ -119,22 +175,34 @@ public Task<(object Snapshot, long Index)> GetSnapshotAsync(string actorName)
...
-                    await db.Events.AddAsync(item);
+                var item = new Event(actorName, index, JsonConvert.SerializeObject(@event, new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.All }));
 
-                    await db.SaveChangesAsync();
+                using (var connection = new SqliteConnection(_connectionString))
+                {
+                    connection.Open();
 
-                    return index++;
+                    using (var transaction = connection.BeginTransaction())
```
Comment (adamhathcock) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Don't need a transaction over a single command.  Sqlite implicitly has a transaction on this.

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 6]
```diff
@@ -147,13 +215,25 @@ public Task<(object Snapshot, long Index)> GetSnapshotAsync(string actorName)
...
-                    var item = new Snapshot(actorName, index, JsonConvert.SerializeObject(snapshot, new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.All }));
+                var item = new Snapshot(actorName, index, JsonConvert.SerializeObject(snapshot, new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.All }));
 
-                    await db.Snapshots.AddAsync(item);
+                using (var connection = new SqliteConnection(_connectionString))
+                {
+                    connection.Open();
 
-                    await db.SaveChangesAsync();
+                    using (var transaction = connection.BeginTransaction())
```
Comment (adamhathcock) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Don't need a transaction over a single command.  Sqlite implicitly has a transaction on this.

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 7]
```diff
@@ -93,24 +139,34 @@ public SqliteProvider(string datasource = "states.db")
...
+                using (var connection = new SqliteConnection(_connectionString))
                 {
-                    var item = db.Snapshots
-                                    .Where(x => x.ActorName == actorName)
-                                    .OrderByDescending(x => x.SnapshotIndex)
-                                    .FirstOrDefault();
+                    connection.Open();
 
-                    if (item != null)
+                    using (var transaction = connection.BeginTransaction())
```
Comment (adamhathcock) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Don't need a transaction over a single cursor read.

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs (hunk 1 above)
Comment (raskolnikoov) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Fixed

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs (hunk 3 above)
Comment (raskolnikoov) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Fixed

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs (hunk 4 above)
Comment (raskolnikoov) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Fixed

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs (hunk 2 above)
Comment (raskolnikoov) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Fixed

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs (hunk 7 above)
Comment (raskolnikoov) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Fixed

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs (hunk 5 above)
Comment (raskolnikoov) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Fixed

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs (hunk 6 above)
Comment (raskolnikoov) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Fixed

**Comment by bricelam** [comment]
I'm curious what the motivation behind this was. Less dependencies?

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 8]
```diff
@@ -29,19 +41,18 @@ public SqliteProvider(string datasource = "states.db")
...
-                using (var db = new SqlitePersistenceContext(_datasource))
+                using (var connection = new SqliteConnection(_connectionString))
                 {
-                    var items = db.Events
-                                    .Where(x => x.ActorName == actorName)
-                                    .Where(x => x.EventIndex <= inclusiveToIndex)
-                                    .ToList();
-
-                    db.RemoveRange(items);
+                    await connection.OpenAsync();
```
Comment (bricelam) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: SQLite doesn't actually support async. It'll be faster if you just use Open.

File: src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs [hunk 9]
```diff
@@ -29,19 +41,18 @@ public SqliteProvider(string datasource = "states.db")
...
 
-                    await db.SaveChangesAsync();
+                    var deleteCommand = connection.CreateCommand();
+                    deleteCommand.CommandText = "DELETE FROM Events WHERE ActorName = $actorName AND EventIndex <= $inclusiveToIndex";
+                    deleteCommand.Parameters.AddWithValue("$actorName", actorName);
+                    deleteCommand.Parameters.AddWithValue("$inclusiveToIndex", inclusiveToIndex);
+                    await deleteCommand.ExecuteNonQueryAsync();
```
Comment (bricelam) [inline src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: Ditto

Comment (bricelam) [reply src/PersistenceProviders/Proto.Persistence.Sqlite/SqliteProvider.cs]: (for all async calls)

**Comment by raskolnikoov** [comment]
@bricelam yes less dependencies :)

**Pull request merged**